// elgato-stream-deck/dist/models/mini.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.StreamDeckMini = void 0;
const util_1 = require("../util");
const base_1 = require("./base");
const id_1 = require("./id");
const miniProperties = {
    MODEL: id_1.DeviceModelId.MINI,
    COLUMNS: 3,
    ROWS: 2,
    ICON_SIZE: 80,
    KEY_DIRECTION: 'ltr',
    KEY_DATA_OFFSET: 1,
};
class StreamDeckMini extends base_1.StreamDeckBase {
    constructor(deviceInfo, options) {
        super(deviceInfo, options, miniProperties);
    }
    transformKeyIndex(keyIndex) {
        return keyIndex;
    }
    convertFillImage(sourceBuffer, sourceOptions) {
        const byteBuffer = util_1.imageToByteArray(sourceBuffer, sourceOptions, util_1.BMP_HEADER_LENGTH, this.rotateCoordinates.bind(this), 'bgr', this.ICON_SIZE);
        util_1.writeBMPHeader(byteBuffer, this.ICON_SIZE, this.ICON_BYTES, 2835);
        return byteBuffer;
    }
    getFillImagePacketLength() {
        return 1024;
    }
    rotateCoordinates(x, y) {
        return { x: this.ICON_SIZE - y - 1, y: x };
    }
}
exports.StreamDeckMini = StreamDeckMini;
